miasm/core/cpu.py

```diff
@@ -1151,13 +1152,52 @@ def add_pre_dis_info(self, prefix=None):
         return True
 
     @classmethod
-    def getbits(cls, bs, attrib, offset, offset_bit, l):
-        return bs.get_bits(offset, offset_bit, l)
+    def getbits(cls, bs, attrib, offset, offset_bit, size):
+        if not size:
+            return 0
+
+        first_byte = offset + (offset_bit // 8)
+        first_bit_in_first_byte = offset_bit % 8
+        number_bytes_to_read = ceil((size + first_bit_in_first_byte) / 8)
+
+        # Read all bytes in a single step
+        endian_offset = cls.endian_offset(attrib, first_byte)
+        endian_offset = endian_offset if attrib == "b" else endian_offset - number_bytes_to_read + 1
+        data = cls.getbytes(bs, endian_offset, number_bytes_to_read)
+        number = cls.convert_endian(attrib, data)
+
+        # Remove the leading bits that we don't care about (if any)
+        if first_bit_in_first_byte != 0:
+            mask = (1 << ((number_bytes_to_read * 8) - first_bit_in_first_byte)) - 1
+            number &= mask
+
+        # Remove the trailing bits that we don't care about (if any)
+        number >>= (8 - (size % 8) - first_bit_in_first_byte) % 8
+
+        return number
 
     @classmethod
     def getbytes(cls, bs, offset, l):
         return bs.get_bytes_exact(offset, l)
 
+    @classmethod
+    def endian_offset(cls, attrib, offset):
```

Comment:
as you are in cpu.py, maybe this fuction should be named "endian_offset_u32" or something approaching?

Reply:
Done